### PR TITLE
Adgeneration: Device info in headers and query params

### PR DESCRIPTION
**Pull request by marki1an**

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/adgeneration/AdgenerationBidder.java [hunk 1]
```diff
@@ -94,7 +94,7 @@ private ExtImpAdgeneration parseAndValidateImpExt(Imp imp) {
         return extImpAdgeneration;
     }
 
-    private String getUri(String adSize, String id, String currency, Site site, Source source) {
+    private String getUri(String adSize, String id, String currency, final BidRequest bidRequest) {
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/adgeneration/AdgenerationBidder.java:97]: Rewrite method like this:
```
private String getUri(String adSize, String id, String currency, final BidRequest bidRequest) {
    final URIBuilder uriBuilder;
    try {
        uriBuilder = new URIBuilder(endpointUrl);
    } catch (URISyntaxException e) {
        throw new PreBidException(String.format("Invalid url: %s, error: %s", endpointUrl, e.getMessage()));
    }

    uriBuilder
            .addParameter("posall", "SSPLOC")
            .addParameter("id", id)
            .addParameter("hb", "true")
            .addParameter("t", "json3")
            .addParameter("currency", currency)
            .addParameter("sdkname", "prebidserver")
            .addParameter("adapterver", VERSION);

    addUriParameterIfNotEmpty(uriBuilder, "sizes", adSize);
    addUriParameterIfNotEmpty(uriBuilder, "tp", ObjectUtil.getIfNotNull(bidRequest.getSite(), Site::getPage));
    addUriParameterIfNotEmpty(uriBuilder, "appbundle", ObjectUtil.getIfNotNull(bidRequest.getApp(), App::getBundle));
    addUriParameterIfNotEmpty(uriBuilder, "appname", ObjectUtil.getIfNotNull(bidRequest.getApp(), App::getName));
    addUriParameterIfNotEmpty(
            uriBuilder, "transactionid", ObjectUtil.getIfNotNull(bidRequest.getSource(), Source::getTid));

    final Device device = bidRequest.getDevice();
    final String deviceOs = device != null ? device.getOs() : null;
    if ("android".equals(deviceOs)) {
        uriBuilder.addParameter("sdktype", "1");
        addUriParameterIfNotEmpty(uriBuilder, "advertising_id", device.getIfa());
    } else if ("ios".equals(deviceOs)) {
        uriBuilder.addParameter("sdktype", "2");
        addUriParameterIfNotEmpty(uriBuilder, "idfa", device.getIfa());
    } else {
        uriBuilder.addParameter("sdktype", "0");
    }

    return uriBuilder.toString();
}

private static void addUriParameterIfNotEmpty(URIBuilder uriBuilder, String parameter, String value) {
    if (StringUtils.isNotEmpty(value)) {
        uriBuilder.addParameter(parameter, value);
    }
}
```

File: src/test/java/org/prebid/server/bidder/adgeneration/AdgenerationBidderTest.java [hunk 2]
```diff
@@ -80,6 +84,28 @@ public void makeHttpRequestsShouldReturnErrorIfIdIsEmpty() {
                 .containsOnly(BidderError.badInput("No Location ID in ExtImpAdgeneration."));
     }
 
+    @Test
+    public void makeHttpRequestsShouldReturnOneRequestPerImpWithExpectedHeaders() {
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/adgeneration/AdgenerationBidderTest.java:88]: Rename to `makeHttpRequestsShouldReturnRequestWithExpectedHeaders`

File: src/test/java/org/prebid/server/bidder/adgeneration/AdgenerationBidderTest.java [hunk 3]
```diff
@@ -80,6 +84,28 @@ public void makeHttpRequestsShouldReturnErrorIfIdIsEmpty() {
...
+        // then
+        assertThat(result.getValue()).hasSize(1)
+                .flatExtracting(r -> r.getHeaders().entries())
+                .extracting(Map.Entry::getKey, Map.Entry::getValue)
+                .containsOnly(
+                        tuple("Content-Type", "application/json;charset=utf-8"),
+                        tuple("Accept", "application/json"),
+                        tuple("User-Agent", "test"),
+                        tuple("X-Forwarded-For", "123"));
+    }
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/adgeneration/AdgenerationBidderTest.java:107]: Rid of `hasSize` and use `containsExactlyInAnyOrder`

File: src/test/java/org/prebid/server/bidder/adgeneration/AdgenerationBidderTest.java [hunk 4]
```diff
@@ -80,6 +84,28 @@ public void makeHttpRequestsShouldReturnErrorIfIdIsEmpty() {
...
+                .imp(singletonList(givenImp(identity())))
+                .device(Device.builder().ip("123").ua("test").build())
+                .build();
+
+        // when
+        final Result<List<HttpRequest<Void>>> result = adgenerationBidder.makeHttpRequests(bidRequest);
+
+        // then
+        assertThat(result.getValue()).hasSize(1)
+                .flatExtracting(r -> r.getHeaders().entries())
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/adgeneration/AdgenerationBidderTest.java:100]: Replace with
```
.extracting(HttpRequest::getHeaders)
.flatExtracting(MultiMap::entries)
```

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/adgeneration/AdgenerationBidder.java (hunk 1 above)
Comment (marki1an) [reply src/main/java/org/prebid/server/bidder/adgeneration/AdgenerationBidder.java:97]: In `go` each parameter goes in a chain, it maybe need to make the same. May effect on URI sequence or correct me.

Comment (CTMBNara) [reply src/main/java/org/prebid/server/bidder/adgeneration/AdgenerationBidder.java:97]: It does not matter what order the query parameters are in.

**Review by CTMBNara — APPROVED** [review]

**Review by SerhiiNahornyi — APPROVED** [review]

**Pull request merged**